cpp/include/cudf/filling.hpp

```diff
@@ -0,0 +1,127 @@
+ * N = (@p end - @p begin).
+ *
+ * The datatypes of @p destination and @p value must be the same. This function
+ * assumes that no memory reallocation is necessary for @p destination. This
+ * function updates in-place and throws an exception if memory reallocation is
+ * necessary (e.g. for variable width types). Use the out-of-place fill function
+ * returning std::unique_ptr<column> for use cases requiring memory
+ * reallocation.
+ *
+ * @param destination The preallocated column to fill into
+ * @param begin The starting index of the fill range
+ * @param end The index one past the end of the fill range
+ * @param value The scalar value to fill
+ * @return void
+ *---------------------------------------------------------------------------**/
+void fill(mutable_column_view& destination, size_type begin, size_type end,
+          gdf_scalar const& value);
+
+/**---------------------------------------------------------------------------*
+ * @brief Fills a range of elements in a column with a scalar value.
+ * 
+ * This fill function updates out-of-place creating a new column object to
+ * return. The returned column holds @p value for N elements from @p begin,
+ * where N = (@p end - @p begin). The returned column stores same values to
+ * @p input outside the fill range.
+ *
+ * The datatypes of @p input and @p value must be the same.
+ *
+ * @param input The input column used to create a new column. The new column
+ * is created by replacing the values of @p input in the specified range with
+ * @p value.
+ * @param begin The starting index of the fill range
+ * @param end The index one past the end of the fill range
+ * @param value The scalar value to fill
+ * @param mr Memory resource to allocate the result output column
+ * @return std::unique_ptr<column> The result output column
+ *---------------------------------------------------------------------------**/
+std::unique_ptr<column> fill(
+    column_view const& input, size_type begin, size_type end,
+    gdf_scalar const& value,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**---------------------------------------------------------------------------*
+ * @brief Repeat rows of a Table.
+ * 
+ * Creates a new table by repeating the rows of @p input_table. The number of 
+ * repetitions of each element is defined by the value at the corresponding 
+ * index of @p count
+ * Example:
+ * ```
+ * in = [4,5,6]
+ * count = [1,2,3]
+ * return = [4,5,5,6,6,6]
+ * ```
+ * @p count should be non-nullable; should not contain negative values; and the
+ * sum of count elements should not overflow the size_type's limit. It is
+ * undefined behavior if @p count has negative values or the sum overflows and
+ * @p check_count is set to false.
+ *
+ * @param input_table Input table
+ * @param count Non-nullable column of a integral type
+ * @param check_count Whether to check count (negative values and overflow)
```

Comment:
Do we need this? Did the old API have this?

Reply:
> Did the old API have this?

No.

> Do we need this?

So, the input parameter column_view const& _count_ should not have negative values and the sum of all the count values in _count_ should not overflow (exceed the limit of size_type). The repeat algorithm first runs thrust::inclusive_scan on the count data and (assuming that inclusive_scan produces a non-decreasing sequence of numbers) run thrust::upper_bound. Thrust does not explicitly specify what happens if the thrust::upper_bound function's precondition is violated (the input sequence for binary search needs to be at least partially sorted), but this may lead to undefined behavior unless thrust explicitly checks this and throws an exception.

Whether should we check this or not is an API design issue, but I saw

https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/include/cudf/legacy/copying.hpp#L307

The gather function has _check_bounds_ (which is set to false by default but gives users an option to be safe) so I thought it may be appropriate to follow this approach. Let me know how you think.

Reply:
How do you plan to check for overflow?

Reply:
```
  rmm::device_vector<size_type> offsets(input_table.num_rows());
  thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
                         count.begin<size_t>(), count.end<size_t>(), offsets.begin());
  if (check_count == true) {
    CUDF_EXPECTS(thrust::is_sorted(rmm::exec_policy(stream)->on(stream),
                                   offsets.begin(), offsets.end()) == true,
                 "count has negative values or the resulting table has more \
                  rows than size_type's limit.");
  }
```

If there is a negative value or overflow, the result of inclusive_scan will not be sorted, so thrust::is_sorted call will return false. This is linear to the number of rows in a table (or a repeat column). Not terribly expensive, but not a constant time. And if this is checked at python level, no need to check this twice, so I introduced check_count, but let me know your thought.

Reply:
I think this is reasonable. My only question is whether the default should be safety (check) or performance (no check).